crates/libs/windows/src/core/event.rs

```diff
@@ -0,0 +1,254 @@
+                }
+            }
+            if removed {
+                let swap_lock = self.swap.lock().unwrap();
+                _lock_free_drop = self.delegates.swap(new_delegates);
+            }
+        }
+        Ok(())
+    }
+    /// Clears the event, removing all delegates.
+    pub fn clear(&mut self) {
+        let mut _lock_free_drop = Array::new();
+        {
+            let change_lock = self.change.lock().unwrap();
+            if self.delegates.is_empty() {
+                return;
+            }
+            let swap_lock = self.swap.lock().unwrap();
+            _lock_free_drop = self.delegates.swap(Array::new());
+        }
+    }
+    /// Invokes all of the event object's registered delegates with the provided callback.
+    pub fn call<F: FnMut(&T) -> Result<()>>(&mut self, mut callback: F) -> Result<()> {
+        let lock_free_calls = {
+            let swap_lock = self.swap.lock().unwrap();
+            self.delegates.clone()
+        };
+        for delegate in lock_free_calls.as_slice() {
+            if let Err(error) = delegate.call(&mut callback) {
+                const RPC_E_SERVER_UNAVAILABLE: HRESULT = HRESULT(-2147023174); // HRESULT_FROM_WIN32(RPC_S_SERVER_UNAVAILABLE)
+                if matches!(error.code(), RPC_E_DISCONNECTED | JSCRIPT_E_CANTEXECUTE | RPC_E_SERVER_UNAVAILABLE) {
+                    self.remove(delegate.to_token())?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// A thread-safe reference-counted array of delegates.
+struct Array<T: Interface + Clone> {
+    buffer: *mut Buffer,
+    len: usize,
+    _phantom: std::marker::PhantomData<T>,
+}
+
+impl<T: Interface + Clone> Default for Array<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: Interface + Clone> Array<T> {
+    /// Creates a new, empty `Array<T>` with no capacity.
+    fn new() -> Self {
+        Self { buffer: std::ptr::null_mut(), len: 0, _phantom: std::marker::PhantomData }
+    }
+    /// Creates a new, empty `Array<T>` with the specified capacity.
+    fn with_capacity(capacity: usize) -> Result<Self> {
+        Ok(Self { buffer: Buffer::new(capacity * std::mem::size_of::<Delegate<T>>())?, len: 0, _phantom: std::marker::PhantomData })
+    }
+    /// Swaps the contents of two `Array<T>` objects.
+    fn swap(&mut self, mut other: Self) -> Self {
+        unsafe { std::ptr::swap(&mut self.buffer, &mut other.buffer) };
+        std::mem::swap(&mut self.len, &mut other.len);
+        other
+    }
+    /// Returns `true` if the array contains no delegates.
+    fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+    /// Returns the number of delegates in the array.
+    fn len(&self) -> usize {
+        self.len
+    }
+    /// Appends a delegate to the back of the array.
+    fn push(&mut self, delegate: Delegate<T>) {
+        unsafe {
+            std::ptr::write((*self.buffer).as_mut_ptr::<Delegate<T>>().add(self.len) as _, delegate);
```

Comment:
What happens if we push beyond the capacity of the buffer?

Reply:
Bad things.